src/pubd/repository.rs

```diff
@@ -764,20 +767,172 @@ pub struct RrdpServer {
+    /// is assumed to be verified and can be applied to the current snapshot.
+    ///
+    /// The new StagedElements is assumed to be verified against the current
+    /// snapshot after *this* existing StagedElements has been applied.
+    ///
+    /// The reason for this merging is that while publishing CAs will always
+    /// see all their objects as reflected by the latest snapshot with any
+    /// existing, but unpublished, updates applied to it. So, they may wish
+    /// to make further updates before the staged updates are actually published.
+    ///
+    /// But if they do, then the new elements should be merged into this existing
+    /// StagedElements so that it can be applied to the current snapshot. This
+    /// way we can have a single RRDP delta that combines all changes.
+    ///
+    /// This operation is not entirely trivial as we need to make sure that
+    /// any hashes in updates and withdraws after merge match the current
+    /// snapshot, i.e. the new snapshot could refer to hashes of not yet
+    /// published objects. We may also simply "forget" objects that were
+    /// staged for publication, and then withdrawn, without ever have been
+    /// published.
+    ///
+    /// Also note (again) that all changes have been verified before when
+    /// this is called. That means that while certain corner cases could be
+    /// problematic (e.g. double withdraw of an object), we will largely
+    /// ignore such issues here. We need to do this, because we get these
+    /// changes as write-ahead-log (WAL) changes and therefore applying
+    /// them is not allowed to fail.
+    fn merge_new_elements(&mut self, elements: DeltaElements) {
+        let (publishes, updates, withdraws) = elements.unpack();
+
+        for pbl in publishes {
+            let uri = pbl.uri().clone();
+            match self.0.get_mut(&uri) {
+                Some(DeltaElement::Publish(_)) => {
+                    // This should never happen as the new publish would have
+                    // been an update instead. This will have been checked
+                    // at this point. Still, rather than failing which is
+                    // not allowed in this code path, we can use the publish
+                    // instead of the staged publish.
+                    self.0.insert(uri, DeltaElement::Publish(pbl));
+                }
+
+                Some(DeltaElement::Update(staged_update)) => {
+                    // This should never happen as the new publish would have
+                    // been an update instead. This will have been checked
+                    // at this point. Still, rather than failing which is
+                    // not allowed in this code path, we can update the
+                    // existing update with the content of the new publish
+                    // element.
+                    staged_update.with_updated_content(staged_update.base64().clone());
```

Comment:
This looks very strange. It looks like it says set A to A... shouldn't it use `pbl` here?

Reply:
indeed. fix pushed